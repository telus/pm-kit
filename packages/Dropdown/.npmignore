*.stories.{js,mdx}
rollup.config.js
Dropdown.jsx
__tests__/