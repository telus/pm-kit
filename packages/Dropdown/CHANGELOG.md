# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.1](https://github.com/telus/pm-kit/compare/@pm-kit/dropdown@1.0.0...@pm-kit/dropdown@1.0.1) (2020-01-24)

**Note:** Version bump only for package @pm-kit/dropdown





# 1.0.0 (2020-01-22)


### Features

* **dropdown:** added Dropdown package ([#31](https://github.com/telus/pm-kit/issues/31)) ([0741099](https://github.com/telus/pm-kit/commit/0741099a74c59f5e0ba6d442b02f67eb22275bfd))
