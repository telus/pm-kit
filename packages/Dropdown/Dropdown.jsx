import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import { css } from '@emotion/core'
import Select, { components, createFilter } from 'react-select'
import { motion, AnimatePresence } from 'framer-motion'
import { size, weight } from '@pm-kit/typography'
import { parkGreen, red, lightTan, white } from '@pm-kit/colours'
import generateId from '../../shared/utils/generateId/generateId.js'

const dropdownWrapper = css`
  width: 100%;
`

const basicLabel = css`
  display: block;
  margin: 0 0 0.2rem 1rem;
  letter-spacing: 0.05px;
  font-size: ${size.bodySmall};
  font-weight: ${weight.normal};
`

const largeLabelStyle = css`
  margin: 0 0 0.5rem 0.2rem;
  font-size: ${size.bodyLarge};
  font-weight: ${weight.bold};
`

const errorFeedback = css`
  margin-left: 0.5rem;
  font-weight: ${weight.normal};
  color: ${red};
  & span {
    color: ${red};
  }
`

const menuStyling = css`
  margin-top: 4px;
`
const menuList = css`
  overflow: hidden;
`

const placeholder = css`
  color: ${parkGreen};
  font-size: 16px;
`

const dropdownAnimate = {
  hide: {
    height: 0,
    transition: { height: { type: 'tween', duration: 0.2 } },
  },
  show: {
    height: 'auto',
    transition: { height: { type: 'tween', duration: 0.2 } },
  },
}

export const Dropdown = ({
  error,
  feedback,
  id,
  label,
  labelType,
  onChange,
  options,
  required,
  type,
  value,
  placeholder,
  ignoreCase,
  ignoreAccents,
  trim,
  matchFrom,
  styles,
  forwardedRef,
  ...rest
}) => {
  const customDropdownStyles = {
    option: (provided, state) => ({
      ...provided,
      padding: '10px 17px',
      borderBottom: '1px dotted #DDD',
      backgroundColor: state.isFocused ? `${lightTan}` : 'inherit',
      color: `${parkGreen}`,
      fontWeight: state.isSelected ? `${weight.bold}` : `${weight.normal}`,
      fontSize: '1rem',
      cursor: 'pointer',
      '&:hover': {
        backgroundColor: `${lightTan}`,
      },
    }),
    control: () => ({
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'space-between',
      minWidth: '100%',
      height: '48px',
      padding: '0 0 0 7px',
      fontSize: '18px',
      borderRadius: '8px',
      border: `1px solid`,
      borderColor: `${feedback === 'error' ? red : 'initial'}`,
      backgroundColor: `${white}`,
      color: `${parkGreen}`,
      cursor: 'pointer',
    }),
    singleValue: () => ({
      color: `${parkGreen}`,
    }),
    indicatorSeparator: provided => ({
      ...provided,
      visibility: 'hidden',
    }),
    dropdownIndicator: provided => ({ ...provided, color: `${parkGreen}` }),
    clearIndicator: () => ({
      display: 'none',
    }),
    placeholder: provided => ({
      ...provided,
      color: `${parkGreen}`,
    }),
  }

  const selectId = generateId(id, label)
  const filterConfig = {
    ignoreCase,
    ignoreAccents,
    trim,
    matchFrom,
  }
  const labelStyleArr = [basicLabel]
  const dropdownWrapperArr = [dropdownWrapper]

  if (labelType === 'large') {
    labelStyleArr.push(largeLabelStyle)
  }

  if (styles) {
    if (styles.dropdownStyle) {
      dropdownWrapperArr.push(styles.dropdownStyle)
    }
    if (styles.labelStyle) {
      labelStyleArr.push(styles.labelStyle)
    }
  }

  return (
    <div css={dropdownWrapperArr}>
      {labelType !== 'hidden' && (
        <label css={labelStyleArr} htmlFor={selectId.identity()}>
          <span>
            {label} {required && '*'}
          </span>
          {feedback === 'error' && <span css={errorFeedback}>({error})</span>}
        </label>
      )}
      <Select
        value={value}
        onChange={onChange}
        options={options}
        placeholder={placeholder}
        id={selectId.identity()}
        type={type}
        styles={customDropdownStyles}
        isClearable={true}
        clearIndicator={false}
        components={{ Menu, Option, SelectContainer, Placeholder }}
        filterOption={createFilter(filterConfig)}
        ref={forwardedRef}
        {...rest}
      />
    </div>
  )
}

const Menu = ({ children, ...props }) => {
  return (
    <components.Menu {...props} css={menuStyling}>
      <motion.div variants={dropdownAnimate} css={menuList} initial="hide" animate="show" exit="hide">
        {children}
      </motion.div>
    </components.Menu>
  )
}

const Option = ({ children, ...props }) => {
  return <components.Option {...props}>{children}</components.Option>
}

const SelectContainer = ({ children, ...props }) => {
  /**
   * Small bug fix that removes the Menu element completley from the DOM when container is out of focus.
   * This is dependent on ReactSelect and AnimatePresence
   */

  //Initially all children including Control and MenuPlacer
  const nonMenuChildren = [...children]
  const menuChildIndex = children.findIndex(child => {
    if (child && child.type.name === 'MenuPlacer') {
      return true
    }
    return false
  })

  //Separates MenuPlacer from the rest of the container Children
  let menuChild = null
  if (menuChildIndex >= 0) {
    menuChild = nonMenuChildren.splice(menuChildIndex, 1)
  }

  //Renders all children but only applies Animation to the MenuPlacer
  return (
    <components.SelectContainer {...props}>
      {nonMenuChildren}
      <AnimatePresence>{menuChild}</AnimatePresence>
    </components.SelectContainer>
  )
}

const Placeholder = props => {
  return <components.Placeholder css={placeholder} {...props} />
}

Dropdown.propTypes = {
  /**
   * An error message. Should be limited to text and links. See usage criteria for more details.
   */
  error: PropTypes.node,
  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(['success', 'error']),
  /**
   * A unique id.
   */
  id: PropTypes.string,
  /**
   * The label.
   */
  label: PropTypes.string.isRequired,
  /**
   * A function called on picking one of the options.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Dropdown options data
   */
  options: PropTypes.array,
  required: PropTypes.bool,
  /**
   * The HTML5 type of the input field.
   */
  type: PropTypes.oneOf(['text', 'number', 'password', 'email', 'search', 'tel', 'url']),
  /**
   * Use `value` for controlled Inputs. For uncontrolled Inputs, use React's built-in `defaultValue` prop.
   * See examples below for more details.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  /**
   * The order of search. 'start' searches options that begin with your search input and
   * 'any' searches options that contain your search input.
   */
  matchFrom: PropTypes.oneOf(['any', 'start']),
  /**
   * Trim your search input.
   */
  trim: PropTypes.bool,
  /**
   * Ignore accents.
   */
  ignoreAccents: PropTypes.bool,
  /**
   * Ignore Case of your search
   */
  ignoreCase: PropTypes.bool,
  /**
   * Displays large label if true is passed. Otherwise, displays regular sized label.
   */
  labelType: PropTypes.oneOf(['large', 'small', 'hidden']),
  /**
   * Customizes the input according to your needs.
   * Accepts an object of styles in the structure below.
   * {
   *  dropdownStyle:{},
   *  labelStyle:{},
   * }
   */
  styles: PropTypes.object,
}

Dropdown.defaultProps = {
  error: undefined,
  feedback: undefined,
  id: undefined,
  options: [],
  required: false,
  type: 'text',
  value: undefined,
  matchFrom: 'start',
  trim: true,
  ignoreAccents: true,
  ignoreCase: true,
  labelType: 'small',
  styles: {},
}

const DropdownWithRef = forwardRef((props, ref) => <Dropdown {...props} forwardedRef={ref} />)

export default DropdownWithRef
