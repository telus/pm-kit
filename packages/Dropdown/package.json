{
  "name": "@pm-kit/dropdown",
  "version": "2.2.10",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "author": "Public Mobile",
  "engines": {
    "node": ">=12"
  },
  "bugs": {
    "url": "https://github.com/telus/public-mobile-design-system/issues"
  },
  "peerDependencies": {
    "@emotion/core": "^11.0.0",
    "react": "^16.12.0",
    "react-dom": "^16.12.0"
  },
  "dependencies": {
    "@pm-kit/colours": "^1.1.6",
    "@pm-kit/feedback-icon": "^2.0.7",
    "@pm-kit/typography": "^1.0.5",
    "framer-motion": "^1.8.3",
    "prop-types": "^15.7.2",
    "react-select": "^3.0.8"
  }
}
