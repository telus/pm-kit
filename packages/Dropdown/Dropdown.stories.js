import React, { useState } from 'react'
import { withKnobs, text, select, object } from '@storybook/addon-knobs'
import Dropdown from './Dropdown'
export default {
  title: 'Design System|Dropdown',
  component: Dropdown,
  decorators: [withKnobs],
}

const cities = [
  { label: 'Toronto', value: '1' },
  { label: 'Ottawa', value: '2' },
  { label: 'Halifax', value: '3' },
  { label: 'Niafara Falls', value: '4' },
  { label: 'Montreal', value: '5' },
  { label: 'Quebec City', value: '6' },
  { label: 'Calgary', value: '7' },
]

export const Default = () => {
  const [city, setCity] = useState(undefined)

  const onCitySelect = citySelected => {
    setCity(citySelected)
  }
  return <Dropdown label="Select a city" value={city} options={cities} onChange={onCitySelect} required={true} />
}

export const LargeLabel = () => {
  const [city, setCity] = useState(cities[0])

  const onCitySelect = citySelected => {
    setCity(citySelected)
  }
  return (
    <Dropdown
      label="Select a city"
      value={city}
      options={cities}
      onChange={onCitySelect}
      required={true}
      labelType="large"
    />
  )
}

export const HideLabel = () => {
  const [city, setCity] = useState(cities[0])

  const onCitySelect = citySelected => {
    setCity(citySelected)
  }
  return (
    <Dropdown
      label="Select a city"
      value={city}
      options={cities}
      onChange={onCitySelect}
      required={true}
      labelType="hidden"
    />
  )
}

export const Error = () => {
  const [errorState, setErrorState] = useState('error')
  const onCitySelect = citySelected => {
    setErrorState('success')
  }
  return (
    <Dropdown
      label="Select a city"
      placeholder="cities"
      options={cities}
      onChange={onCitySelect}
      required={true}
      big={true}
      feedback={errorState}
      error="City is required"
    />
  )
}

export const Playground = () => {
  return (
    <Dropdown
      label={text('Label', 'Testing')}
      required={select('Required', [true, false], false)}
      feedback={select('feedback', ['error', 'success'], 'success')}
      error={text('Error', 'This field is required')}
      options={cities}
      onChange={() => {}}
      placeholder={text('placeholder', 'Cities')}
      matchForm={select('match form', ['start', 'any'], 'start')}
      trim={select('trim', [true, false], true)}
      ignoreAccents={select('ignore accents', [true, false], true)}
      ignoreCase={select('ignore case', [true, false], true)}
      labelType={select('label type', ['hidden', 'large', 'small'], 'small')}
      styles={object('Styles', {
        dropdownStyle: {},
        labelStyle: {},
      })}
    />
  )
}

Playground.story = {
  name: 'playground',
  parameters: { docs: { page: null, disable: true } },
}
