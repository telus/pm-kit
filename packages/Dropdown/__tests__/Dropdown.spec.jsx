import React from 'react'
import { render } from 'enzyme'

import Dropdown from '../Dropdown'

describe('Dropdown', () => {
  it('renders', () => {
    const dropdown = render(<Dropdown />)

    expect(dropdown).toMatchSnapshot()
  })
})
