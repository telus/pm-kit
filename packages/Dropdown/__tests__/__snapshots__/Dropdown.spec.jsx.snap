// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders default dropdown 1`] = `
.emotion-10 {
  width: 100%;
}

.emotion-0 {
  display: block;
  margin: 0 0 0.2rem 1rem;
  -webkit-letter-spacing: 0.05px;
  -moz-letter-spacing: 0.05px;
  -ms-letter-spacing: 0.05px;
  letter-spacing: 0.05px;
  font-size: 0.875rem;
  font-weight: 300;
}

.emotion-9 {
  position: relative;
  box-sizing: border-box;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-width: 100%;
  height: 48px;
  padding: 0 0 0 7px;
  font-size: 18px;
  border-radius: 8px;
  border: 1px solid;
  border-color: initial;
  background-color: #FFF;
  color: #034045;
  cursor: pointer;
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 2px 8px;
  -webkit-overflow-scrolling: touch;
  position: relative;
  overflow: hidden;
  box-sizing: border-box;
}

.emotion-1 {
  color: #034045;
  margin-left: 2px;
  margin-right: 2px;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  box-sizing: border-box;
  color: #034045;
  font-size: 16px;
}

.emotion-2 {
  margin: 2px;
  padding-bottom: 2px;
  padding-top: 2px;
  visibility: visible;
  color: hsl(0,0%,20%);
  box-sizing: border-box;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
}

.emotion-4 {
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  background-color: hsl(0,0%,80%);
  margin-bottom: 8px;
  margin-top: 8px;
  width: 1px;
  box-sizing: border-box;
  visibility: hidden;
}

.emotion-6 {
  color: #034045;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 8px;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  box-sizing: border-box;
}

.emotion-6:hover {
  color: hsl(0,0%,60%);
}

.emotion-5 {
  display: inline-block;
  fill: currentColor;
  line-height: 1;
  stroke: currentColor;
  stroke-width: 0;
}

<div
  class="emotion-10"
>
  <label
    class="emotion-0"
    for="testing"
  >
    <span>
      testing 
    </span>
  </label>
  <div
    class=" emotion-9"
    id="testing"
  >
    <div
      class=" emotion-8"
    >
      <div
        class=" emotion-3"
      >
        <div
          class="emotion-1"
        >
          Select...
        </div>
        <div
          class="emotion-2"
        >
          <div
            class=""
            style="display:inline-block"
          >
            <input
              aria-autocomplete="list"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              id="react-select-2-input"
              spellcheck="false"
              style="box-sizing:content-box;width:1px;label:input;background:0;border:0;font-size:inherit;opacity:1;outline:0;padding:0;color:inherit"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              style="position:absolute;top:0;left:0;visibility:hidden;height:0;overflow:scroll;white-space:pre"
            />
          </div>
        </div>
      </div>
      <div
        class=" emotion-7"
      >
        <span
          class=" emotion-4"
        />
        <div
          aria-hidden="true"
          class=" emotion-6"
        >
          <svg
            aria-hidden="true"
            class="emotion-5"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
